Feature/feature module rewrite

AnimusImmodicus opened the pull request:
This pull request changes the mechanism by which we collect and communicate Prestashop order information to AvaTax for tax calculation. The old mechanism was both inaccurate (missed many tax edge cases) and inefficient (too many API requests).  It also fixes, several UI issues previously caused by the module.

Mekki78 commented:
Hello, as I can see on your Pull Request you are requesting for major update of your module,

$this->version = '3.5.5' -> $this->version = '3.6.0';

This require to correct the main errors you have on our validator.

Please correct this following sections in priority on your module.
[Requirements, Structure, Errors, Optimizations, Security] 

![download 38](https://cloud.githubusercontent.com/assets/6247743/12017735/de4b709e-ad5a-11e5-8713-5a5f03dfb582.png)

Thanks you in advance.

shaabans commented:
Good point about the version number, it might even make sense to use 4.0.0 since the approach we've taken with the module is so different?

Ok, we'll go through the validator (didn't know it existed) and resubmit shortly.

Mekki78 commented:
Thanks, this is the validator link (if you didn't know it existed) https://validator.prestashop.com/
also I give you the link of an important issue you have on  your module https://github.com/PrestaShop/avalaratax/issues/18 (Hardcoded database prefix)

I await your response.
Regards,

AnimusImmodicus commented:
Hello,

I have fixed [Issue #18](https://github.com/PrestaShop/avalaratax/issues/18).

I have also fixed the majority of the issues from the Prestashop Validator. The *Requirements*, *Structure*, and *Security* sections are completely resolved. I have also resolved almost all of the issues from the *Standards* and *Optimizations* sections.

Additionally, I have resolved as many issues as possible from the *Errors* section. However, the validator still complains about use of `isset()`, `$_POST` and `$_GET`. However, if I replace them with `Tools::getIsset()` and `Tools::getValue()` respectively, then the module no longer functions properly. I investigated this issue and it appears that `Tools::getIsset()` and `Tools::getValue()` both perform additional `empty()` and `!is_string()` checks. Because our module relies on both "falsey" and non-string values, it is impossible to use `Tools::getIsset()` and `Tools::getValue()`.

Please let me know if there are other issues we need to resolve.

Take Care,
-Martin

Mekki78 commented:
#19

AnimusImmodicus commented:
Our latest commit 078f888e2e6f19d2c8909a6bd5c03bf5673e7792 should fix Issue [#19]

Mekki78 commented:
Merge Pull Request [#17]

Have a great day !
